### ISSUE TITLE
Unexpected behavior of MaxConcurrency and MaxBufferedItemCount

### ISSUE THREAD

**Issue by johnstairs** [Azure/azure-cosmos-dotnet-v3#2290]
**Background**

We run a REST API that accepts arbitrary queries (similar to OData) that we translate into Cosmos DB queries. Each response has a maximum page size, after which we provide a link to the next page of results. The Cosmos DB container could have a large number of physical partitions (one customer has 2068). All queries need to be cross-partition.

Some queries will be very selective (resulting in 1 or 2 documents), whereas some are not selective at all (millions of results or more). We don't necessarily know whether the query in a request is going to be selective or not. Our goal is to provide good query latency while minimizing RU consumption, notably we want to avoid wasting RUs for results that we don't end up returning to the client.

**What we are trying to achieve**

Ideally, we would like for the selective queries to be issued across all partitions concurrently (`MaxConcurrency=-1`). The non-selective queries should run against a single partition (`MaxConcurrency=null`) since each physical partition is expected to have plenty of results. But we don't which queries are which, so we'd like to gradually fan-out the until we get enough results. In general, we will not be enumerating a query until the end, because we are only returning one page or results at a time.

I was experimenting with trying increase the the fan-out exponentially. Something like (pseudocode):

``` csharp
int partitionsToQuery = 1;
string continuationToken = null;
List results = new();
do
{
   var it = container.GetItemQueryIterator(query, new { MaxItemCount = desiredPageSize, MaxConcurrency = partitionsToQuery, MaxBufferedItemCount = 0  });
   var response = await it.ReadNextAsync();
   results.AddRange(response);
   partitionsToQuery *= 10; // no point in doing this beyond the # physical partitions, we'd just call it.ReadNextAsync() instead
   continuationToken = response.ContinuationToken;
}
while (results.Count < desiredPageSize && continuationToken !=null);

return (results, continuationToken);
``` 

But this approach will not have the desired effect because of what `MaxItemCount`, `MaxConcurrency`, and `MaxBufferedItemCount` really mean.

**Analysis of `MaxItemCount`, `MaxConcurrency`, and `MaxBufferedItemCount`**

 To better understand what is going on, I wrote the following little program:

``` csharp
using System;
using System.Linq;
using System.Reflection;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Azure.Cosmos;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;

Console.WriteLine($"Using Cosmos DB SDK {typeof(CosmosClient).Assembly.GetCustomAttributes<AssemblyFileVersionAttribute>().Single().Version}");

IConfigurationRoot configuration = new ConfigurationBuilder()
    .AddJsonFile("appSettings.json")
    .Build();

string endpoint = configuration["EndPointUrl"];
string authKey = configuration["AuthorizationKey"];
string databaseId = configuration["DatabaseId"];
string containerId = configuration["ContainerId"];

var client = new CosmosClient(endpoint, authKey, new CosmosClientOptions() { CustomHandlers = { new MyHandler() }, ConnectionMode = ConnectionMode.Direct });
var container = client.GetContainer(databaseId, containerId);

await IssueQuery(
    "select * from c",
    new() { MaxItemCount = 100 });

await IssueQuery(
    "select * from c where c.resourceId = 'b89b512b-1ce1-4da9-9952-c76889ad24c5'",
    new() { MaxItemCount = 100 });

await IssueQuery(
    "select * from c where c.resourceId = 'b89b512b-1ce1-4da9-9952-c76889ad24c5'",
    new() { MaxItemCount = 100, MaxConcurrency = -1 });

await IssueQuery(
    "select * from c",
    new() { MaxItemCount = 100, MaxConcurrency = 1 });

await IssueQuery(
    "select * from c",
    new() { MaxItemCount = 100, MaxConcurrency = 1, MaxBufferedItemCount = 0 });


async Task IssueQuery(string query, QueryRequestOptions queryRequestOptions)
{
    Console.WriteLine("*****");
    Console.WriteLine($"Beginning `{query}` with options `{JsonConvert.SerializeObject(queryRequestOptions, new JsonSerializerSettings() { NullValueHandling = NullValueHandling.Ignore })}`");

    var itemQueryIterator = container.GetItemQueryIterator<object>(query, null, queryRequestOptions);

    // keep iterating until we get MaxItemCount or there are no more results.
    for (int count = 0; itemQueryIterator.HasMoreResults && count < queryRequestOptions.MaxItemCount;)
    {
        var feedResponse = await itemQueryIterator.ReadNextAsync();
        count += feedResponse.Count;

        Console.WriteLine($"\tFeedResponse count: {feedResponse.Count}, has CT: {!string.IsNullOrEmpty(feedResponse.ContinuationToken)}, request charge: {feedResponse.RequestCharge}");
    }

    Console.WriteLine("Done.");
    Console.WriteLine();
}

public class MyHandler : RequestHandler
{
    public override async Task<ResponseMessage> SendAsync(RequestMessage request, CancellationToken cancellationToken)
    {
        var responseMessage = await base.SendAsync(request, cancellationToken);

        Console.WriteLine($"\t\tRequestHandler message sent. Physical partition = { responseMessage.Headers["x-ms-documentdb-partitionkeyrangeid"]}, charge = {responseMessage.Headers.RequestCharge}, document count = {responseMessage.Headers["x-ms-item-count"]}");

        return responseMessage;
    }
}
```

I'm trying out a non-selective query and a selective one over a container with 10 physical partitions, experimenting with different values for `MaxConcurrency` and `MaxBufferedItemCount`. To see the actual requests that go to the DB, there's a request handler that outputs information about the response.

Let's go through the output, query by query.

```
Beginning `select * from c` with options `{"MaxItemCount":100}`
                RequestHandler message sent. Physical partition = 10, charge = 9.51, document count = 100
        FeedResponse count: 100, has CT: True, request charge: 9.51
Done.
```

When we don't specify anything for `MaxConcurrency`, we end up with one call to one physical partition, which returns 100 documents to fill our page. Great.

Now let's try a selective query:
```
Beginning `select * from c where c.resourceId = 'b89b512b-1ce1-4da9-9952-c76889ad24c5'` with options `{"MaxItemCount":100}`
                RequestHandler message sent. Physical partition = 10, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 11, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 14, charge = 2.92, document count = 1
        FeedResponse count: 1, has CT: True, request charge: 2.92
                RequestHandler message sent. Physical partition = 15, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 7, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 4, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 12, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 13, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 8, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: True, request charge: 2.79
                RequestHandler message sent. Physical partition = 9, charge = 2.79, document count = 0
        FeedResponse count: 0, has CT: False, request charge: 2.79
Done.
```

We had enumerate the whole way though the FeedIterator because there was only one document that matched this query. Doing this sequentially for a large number of partitions is not great, so let's try to set `MaxConcurrency` to -1:

```
Beginning `select * from c where c.resourceId = 'b89b512b-1ce1-4da9-9952-c76889ad24c5'` with options `{"MaxItemCount":100,"MaxConcurrency":-1}`
                RequestHandler message sent. Physical partition = 11, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 7, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 4, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 8, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 9, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 10, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 13, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 12, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 15, charge = 2.79, document count = 0
                RequestHandler message sent. Physical partition = 14, charge = 2.92, document count = 1
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 1, has CT: True, request charge: 2.92
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 0, has CT: True, request charge: 2.79
        FeedResponse count: 0, has CT: False, request charge: 2.79
Done.
```

Same number of FeedResponses, but subsequent calls to `ReadNextAsync()` return the buffered results. Cool.

Let's go back to the non-selective query, but this time, let's set `MaxConcurrency` to 1, which I would have guessed would do the same thing as leaving it null:

```
Beginning `select * from c` with options `{"MaxItemCount":100,"MaxConcurrency":1}`
                RequestHandler message sent. Physical partition = 10, charge = 9.51, document count = 100
                RequestHandler message sent. Physical partition = 11, charge = 9.53, document count = 100
                RequestHandler message sent. Physical partition = 14, charge = 9.07, document count = 100
                RequestHandler message sent. Physical partition = 15, charge = 9.61, document count = 100
                RequestHandler message sent. Physical partition = 7, charge = 9.21, document count = 100
                RequestHandler message sent. Physical partition = 4, charge = 9.62, document count = 100
                RequestHandler message sent. Physical partition = 12, charge = 9.67, document count = 100
                RequestHandler message sent. Physical partition = 13, charge = 9.4, document count = 100
                RequestHandler message sent. Physical partition = 8, charge = 9.73, document count = 100
                RequestHandler message sent. Physical partition = 9, charge = 9.88, document count = 100
        FeedResponse count: 100, has CT: True, request charge: 9.51
Done.
```

It does something very different. It sequentially queries each physical partition for 100 items. The first FeedResponse has the 100 items I need so the rest is discarded. I now have spent 10x the RUs than were necessary. For our customer with 2068 physical partitions, I suppose it would be 2068x. It seems like setting `MaxConcurrency` to a non-null value indicates that the SDK is going to prefetch results (and wait for all physical partitions to respond before returning a FeedResponse).

So let's try setting `MaxBufferedItemCount` to 0:
```
Beginning `select * from c` with options `{"MaxBufferedItemCount":0,"MaxItemCount":100,"MaxConcurrency":1}`
                RequestHandler message sent. Physical partition = 10, charge = 9.51, document count = 100
                RequestHandler message sent. Physical partition = 11, charge = 9.53, document count = 100
                RequestHandler message sent. Physical partition = 14, charge = 9.07, document count = 100
                RequestHandler message sent. Physical partition = 15, charge = 9.61, document count = 100
                RequestHandler message sent. Physical partition = 7, charge = 9.21, document count = 100
                RequestHandler message sent. Physical partition = 4, charge = 9.62, document count = 100
                RequestHandler message sent. Physical partition = 12, charge = 9.67, document count = 100
                RequestHandler message sent. Physical partition = 13, charge = 9.4, document count = 100
                RequestHandler message sent. Physical partition = 8, charge = 9.73, document count = 100
                RequestHandler message sent. Physical partition = 9, charge = 9.88, document count = 100
        FeedResponse count: 100, has CT: True, request charge: 9.51
Done.
```

No change. In browsing the source for this repo, I couldn't see where the value was actually being used.

**What we would like**

We would like better control over how queries performed against physical partitions. Querying partitions in parallel should not imply eagerly prefetching beyond `MaxItemCount`.

Really what we are trying to achieve, as mentioned above, is balancing good latency for selective queries across a large number of partitions without wasting RUs for non-selective queries, without knowing how many results a query will actually return ahead of time.

If there were a way to say "I need 100 results", query 10 partitions in parallel, and then give me a continuation token, that would be great.

If there were a way to say "I need 100 results", start querying 1, then 10, then 100 partitions in parallel (or 1, 2, 4, etc. ), and then give me a continuation token, that would be really great.

**Environment summary**
SDK Version: 3.17.0
OS Version: Windows

**Comment by timsander1** [comment]
Hi @johnstairs, thanks for all that detail! You should think of MaxConcurrency as the maximum amount of parallelism that you want to allow for cross-partition queries. It is strictly a maximum though, and there aren't any guarantees beyond that.

For the scenario that you've mentioned, this would require some custom code. What kind of latency are you looking for with this query? In other words, if you were to find a balance between latency and the RU charge, where would it be?

Also, how common is the case with many results vs few? Is it a 50/50 split or skew in one direction?

Let me know and I'll share some ideas. Thanks!

**Comment by johnstairs** [comment]
Hi @timsander1, thanks for your reply. I think there are a couple of issues and gaps here:
- MaxConcurrency also implies prefetching, which is not going to be desirable in all cases (for example, when we get a full page of results from one partition, why keep prefetching from other partitions before returning)
- MaxBufferedItemCount doesn't seem to do anything 
- This really breaks down when you have a large number of partitions
- The API seems to be primarily designed for scenarios where the client will consume all results from the query, not just one page

We currently keep `MaxConcurrency = null`, which works well when there are full pages of results in each or most partitions. But it falls apart when a selective query comes in. For a customer with 2048 partitions, end making 2048 sequential calls to the partitions, which takes 25 seconds but with only an average RU charge of 2.8 per partition. 25 seconds here is definitely not acceptable performance for what is a in this case very simple query. We need this query to run in parallel. (And no, we don't know the partition key)

As for which kind of query is more frequent, it really depends on the customer and workload.

What we plan to do on our end is to build a simple cache of queries that were selective that will be populated as the service runs so that we can know to set `MaxConcurrency = -1` next time we see them.

But what about the first time we see a query? That is what this issue is about. We'd like some mechanism to gradually increase the parallelism of the query. Something like:
- Until you get 100 results:
  - Query one partition
  - Now query the next 10 partitions in parallel.
  - Now query the next 100 partitions in parallel.
  - etc.

We would tune the growth factor (10 might be too much, or maybe it would be calculated based on the number of partitions).

How can we accomplish something like this? Right now, setting `MaxConcurrency=10` will end up querying all partitions, which is not what we want.

**Comment by timsander1** [comment]
Based on the query continuation token, the SDK can tell if the FeedResponse is empty vs expensive query/throttling. This is not easily human readable though and would be tough to take a dependency on in your code.

The workaround you've described (cache the number of results for different queries to decide the MaxConcurrency) seems like it could work but, like you mentioned, would result in high latency the first time you see some queries. For this case, you could start at something like 10 (instead of 1), to improve latency.

To better understand the requirements, could you answer these questions?
  - Why don't you know the partition key? With 2,048+ physical partitions, this is an extremely large workload. Whatever solution we can come up does not solve the core problem of a query needing to fan-out to every partition. You've mentioned that this is not a rare operation, so this is definitely something we want to try to avoid. 
  - What is the query that you'll be running? The above example used "SELECT * FROM c" but I'd imagine it's something different right?
  - You mentioned 25 seconds of latency is too high. What is acceptable latency?

**Comment by brendankowitz** [comment]
@timsander1 there are some queries where the partitionkey is not known. Our data is using https://docs.microsoft.com/en-us/azure/cosmos-db/partitioning-overview#using-item-id-as-the-partition-key as the partitioning strategy, this means direct reads are always fast (when the ID is known). Each item can store history, so this scheme also allows the history to be inserted within a transaction. It also allows data to scale out to many TB with no replication. 

There are some rare queries that are very selective (only a handful of results), i.e. searching for a user-provided "identifier" or "tag" value. There isn't a great lookup of this and it is only evident at large scale where all perf matters. Search parameters are also flexible, so you could use "identifier + tag" or only one (or in combination with others i.e. firstname). We do "prefix" the partitionkey and could guess the first half, but I don't think this is useful.

We'll create a new issue, but related to this, it seems like the SDK doesn't return when max parallelism goes over the range of ~500-700 partitions.
